barter-data/src/exchange/coinbase/book/l2.rs

```diff
@@ -0,0 +1,665 @@
+        };
+
+        match input {
+            CoinbaseOrderBookL2Message::Snapshot(snapshot) => {
+                let snapshot = MarketEvent::from((
+                    Coinbase::ID,
+                    instrument.key.clone(),
+                    snapshot,
+                ));
+
+                vec![Ok(snapshot)]
+            }
+            CoinbaseOrderBookL2Message::Update(update) => {
+                let valid_update = match instrument.sequencer.validate_sequence(update) {
+                    Ok(Some(valid_update)) => valid_update,
+                    Ok(None) => return vec![],
+                    Err(error) => return vec![Err(error)],
+                };
+
+                MarketIter::<InstrumentKey, OrderBookEvent>::from((
+                    Coinbase::ID,
+                    instrument.key.clone(),
+                    valid_update,
+                )).0
+            }
+        }
+    }
+}
+
+impl From<CoinbaseOrderBookL2Change> for Level {
+    fn from(change: CoinbaseOrderBookL2Change) -> Self {
+        Self {
+            price: change.price,
+            amount: change.size,
+        }
+    }
+}
+
+impl FromIterator<CoinbaseOrderBookL2Change> for Vec<Level> {
+    fn from_iter<T: IntoIterator<Item=CoinbaseOrderBookL2Change>>(iter: T) -> Self {
+        iter.into_iter().map(Level::from).collect()
+    }
+}
+
+
+impl<InstrumentKey> From<(ExchangeId, InstrumentKey, CoinbaseOrderBookL2Snapshot)> for
+MarketEvent<InstrumentKey, OrderBookEvent> {
+    fn from(
+        (exchange, instrument, snapshot): (
+            ExchangeId,
+            InstrumentKey,
+            CoinbaseOrderBookL2Snapshot,
+        ),
+    ) -> Self {
+        MarketEvent {
+            time_exchange: Utc::now(),
+            time_received: Utc::now(),
+            exchange,
+            instrument,
+            kind: OrderBookEvent::Snapshot(OrderBook::new(
+                0,
+                Some(Utc::now()),
+                snapshot.bids,
+                snapshot.asks,
+            )),
+        }
+    }
+}
+
+
+impl<InstrumentKey> From<(ExchangeId, InstrumentKey, CoinbaseOrderBookL2Update)>
+for MarketIter<InstrumentKey, OrderBookEvent>
+{
+    fn from(
+        (exchange, instrument, update): (
+            ExchangeId,
+            InstrumentKey,
+            CoinbaseOrderBookL2Update,
+        ),
+    ) -> Self {
+        let (bids, asks): (Vec<_>, Vec<_>) = update.changes.into_iter().partition(|change| change.side == Side::Buy);
+
+        Self(vec![Ok(MarketEvent {
+            time_exchange: update.time_exchange,
+            time_received: Utc::now(),
+            exchange,
+            instrument,
+            kind: OrderBookEvent::Update(OrderBook::new(
+                0,
+                Some(update.time_exchange),
+                bids,
+                asks,
+            )),
+        })])
+    }
+}
+
+fn get_book_snapshot_url(symbol: &str) -> String {
```

Comment:
This appears to only be used inside tests. Let's move that into the `mod tests` module as it's essentially dead code